packages/admin-ui/src/masked-input/use-masked-input.ts

```diff
@@ -0,0 +1,184 @@
+import React, { useMemo } from 'react'
+import { useSafeLayoutEffect } from '@vtex/admin-ui-hooks'
+
+import { useDeleteKeyPressed } from './use-delete-key-pressed'
+import { useRerender } from './use-rerender'
+
+/**
+ * React hook that handles masking of any input given a value and formatter.
+ * @example
+ * const [value, setValue] = React.useState('')
+ *
+ * const formatter = useFormatter({
+ *  mask: '__/__/___',
+ *  accept: /[\d]/gi
+ * })
+ *
+ * const inputProps = useMaskedInput({
+ *  value,
+ *  formatter,
+ *  onChange: setValue,
+ * })
+ *
+ * <input {...inputProps} />
+ */
+export function useMaskedInput(
+  props: UseMaskedInputProps
+): UseMaskedInputReturn {
+  const { formatter, value, onChange, accept = /\d/g } = props
+
+  const valueRef = React.useRef<ValueRef | null>(null)
+  const userValue = useMemo(() => formatter(value), [formatter, value])
+  const isDeleteKeyPressed = useDeleteKeyPressed()
+  const rerender = useRerender()
+
+  const handleChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
+    const eventValue = evt.target.value
+
+    valueRef.current = {
+      eventValue,
+      input: evt.target,
+      hasSizeIncrease: eventValue.length > userValue.length,
+      pressedDelete: isDeleteKeyPressed(),
+      hasNoChanges: userValue === formatter(eventValue),
+    }
+
+    rerender()
+  }
+
+  useSafeLayoutEffect(() => {
+    if (valueRef.current == null) return
+
+    const { eventValue, input, hasSizeIncrease, pressedDelete, hasNoChanges } =
+      valueRef.current
+
+    valueRef.current = null
+
+    const deleteCausesNoChanges = pressedDelete && hasNoChanges
+    const valueAfterSelectionStart = eventValue.slice(
+      input?.selectionStart ?? undefined
+    )
+
+    const acceptedCharIndexAfterDelete = valueAfterSelectionStart.search(accept)
+    const charsToSkipAfterDelete =
+      acceptedCharIndexAfterDelete !== -1 ? acceptedCharIndexAfterDelete : 0
+
+    const valueBeforeSelectionStart = clean(
+      eventValue.substring(0, input?.selectionStart ?? undefined),
+      accept
+    )
+
+    const formattedValue = formatter(eventValue)
+
+    if (userValue === formattedValue) {
+      rerender()
+    } else {
+      onChange(formattedValue)
+    }
+
+    return () => {
+      let start = getCursorPosition({
+        formattedValue,
+        valueBeforeSelectionStart,
+        accept,
+      })
+
+      if (hasSizeIncrease || (pressedDelete && !deleteCausesNoChanges)) {
+        while (
+          formattedValue[start] &&
+          clean(formattedValue[start], accept) === ''
+        ) {
+          start += 1
+        }
+      }
+
+      const positionShift =
+        start + (deleteCausesNoChanges ? 1 + charsToSkipAfterDelete : 0)
+
+      input.selectionEnd = positionShift
+      input.selectionStart = positionShift
+    }
+  })
+
+  return {
+    value: valueRef.current !== null ? valueRef.current.eventValue : userValue,
+    onChange: handleChange,
```

Comment:
I think it would be good to the accessibility to have the input maxLenght defined
```suggestion
    onChange: handleChange,
    maxLength: mask.length
```

Reply:
This hook does not receive a mask property.